Add the ability to parse error objects

gfontenot opened the pull request:
This adds a `transformError` function to `Request`. This acts as a hook
to be able to change the error from the `NSError` used by the lower
level network clients and JSON parsing into a model more suitable for
your application.

By default, we are still going to return `NSError`s, but it's useful to
have a hook for being able to model the network error response if you
know what it looks like.

This maintains 100% backwards compatibility with the existing API.

Fixes #46

gfontenot committed:
Add the ability to parse error objects

This adds a `transformError` function to `Request. This acts as a hook
to be able to change the error from the `NSError` used by the lower
level network clients and JSON parsing into a model more suitable for
your application.

By default, we are still going to return `NSError`s, but it's useful to
have a hook for being able to model the network error response if you
know what it looks like.

This maintains 100% backwards compatibility with the existing API.

kaishin commented:
Interesting approach that didn't occur to me before. One tiny nitpick: success type is called `ResponseType` but error type is called `ResponseError` instead of `ResponseErrorType`. If we could rename things I'd probably go with something like `ResponseSuccessType` and `ResponseFailureType` or something like that.

I'm +1 for merging this as is for now.

gfontenot commented:
Yeah, had the same thought. I was thinking about renaming `ResponseType` to `ResponseObject`.

kaishin commented:
I like `ResponseObject` but wouldn't that break backward compat?

gfontenot commented:
Sure would, that's why I haven't changed it. I think that's an OK change to make before we release 1.0 though.

sidraval commented:
Nice!

gfontenot commented:
Can't decide if this is the right approach. Any more feedback from @thoughtbot/ios?

gfontenot commented:
Also ping @paulyoung in case he has thoughts on this

paulyoung commented:
I think this seems fine. Without trying to do something custom I can't really offer any more feedback right now :+1:

bernerdschaefer commented:
Does `NSError` contain enough information to transform into a meaningful domain error object?

I understood #46 as allowing users to work with structured error responses in APIs, like those [returned by Heroku's Platform API](https://devcenter.heroku.com/articles/platform-api-reference#errors).

gfontenot commented:
@bernerdschaefer Not really. I've actually been working on taking another stab at this. Waiting on some other PRs to come through before I can finish.

gfontenot committed:
Wrap errors in SwishError instead of NSError

Previously, we were passing around instances of NSError internally. This
worked fine, but ultimately doesn't provide enough error info to be able
to be able to do any kind of meaningful transformations down the line.

By wrapping our errors in a SwishError enum, we can be more explicit
about where the errors are coming from, and what they represent.

gfontenot commented:
Finally got this updated so that we're not using a raw `NSError` for the transformation. I'd love some more eyes on this if anyone has time.

sidraval commented:
This looks great to me.

sharplet commented on Source/Models/SwishError.swift:
```diff
@@ -0,0 +1,26 @@
+  case SessionError(NSError)
+  case ServerError(code: Int, json: AnyObject)
+  case JSONSerializationError(NSError)
+  case ArgoError(DecodeError)
```
Do you think it's a smell that every case here ends in `Error`? It feels like duplication to me.

gfontenot replied:
Yeah, for sure. Would you just chop off `Error` at the end of the specific cases?

sharplet replied:
That's a good start. I think it would be good to then look at each name and consider whether it's clear from just the name where the error occurred.

For example: `case Session(NSError)`. Is that wrapping an error returned from an `NSURLSession` task? It's not entirely clear.

gfontenot replied:
Do you have naming suggestions?

sharplet commented on Source/Models/SwishError.swift:
```diff
@@ -0,0 +1,26 @@
+public enum SwishError {
+  case SessionError(NSError)
+  case ServerError(code: Int, json: AnyObject)
+  case JSONSerializationError(NSError)
```
It looks like this case covers both serialization and deserialization. Is there value in separating those out? My first impression of this was to wonder where the deserialization case was, before I read the code and realized it was included here.

gfontenot replied:
I think this is actually _only_ deserialization, not serialization at all. Should probably rename this to make that clear.

sharplet commented on Source/Models/SwishError.swift:
```diff
@@ -0,0 +1,26 @@
+extension SwishError: ErrorType { }
+
+public extension SwishError {
+  var rawError: NSError {
```
What do you think about changing this to `underlyingError: ErrorType`?

gfontenot replied:
I think that probably makes sense. We can then move from `ErrorType -> NSError` for the default `transformError` method. The only thing that wouldn't work for is the `.ServerError` case. Would you still transform that into an `NSError`?

sharplet replied:
Yeah, well I suppose the idea of an "underlying" error doesn't really make sense in that case. The response _is_ the error, there's nothing underlying. I suppose you could make the return type optional? But it depends on how this method is intended to be used.

gfontenot replied:
Primarily, this is used to transform `SwishError` -> `NSError` for the default `transformError` implementation.

sharplet commented:
This looks pretty good, I've just left a few notes about the design of `SwishError`.

tonyd256 commented:
I don't have more to add than Adam. Good change :+1:

sharplet commented on Source/Protocols/Request.swift:
```diff
@@ -6,24 +6,32 @@ public typealias EmptyResponse = Void
   func build() -> NSURLRequest
-  func parse(j: JSON) -> Result<ResponseObject, NSError>
+  func parse(j: JSON) -> Result<ResponseObject, SwishError>
+  func transformError(error: SwishError) -> ResponseError
```
Is the intended use case for this to allow clients to map `SwishError` directly into their own error types? What if this function instead accepted a function `<E: ErrorType> SwishError -> E`?

sharplet replied:
This would essentially turn it into a `mapError` function I think

gfontenot replied:
I'm unclear on where users would be able to actually _use_ that method. It'd move the logic outside the request and into some object up above the `APIClient`.

gfontenot replied:
And yes, that's what this function is intended for. It mimics `parse` but for the error case.

sharplet committed:
Improve clarity of some SwishError case names

sharplet commented:
Rather than continue all the various threads, I thought a diff would be a lot clearer. I've pushed my proposed changes here: https://github.com/thoughtbot/Swish/compare/gf-model-errors...as-alt-swish-error.

What I ended up with was that each error case with an underlying error kept the `Error` suffix. But `JSONSerializationError` was renamed to `InvalidJSONResponse` and `SessionError` to `URLSessionError` for clarity.

I also renamed `rawError` to `NSError` (though I'm not 100% convinced about the capitalization).

gfontenot commented:
I've updated this PR with @sharplet's naming changes. Any other comments?

sharplet commented:
LGTM 👍
